### PR TITLE
refactor: Replace magic numbers with named constants

### PR DESCRIPTION
**Pull request by dssjon**
Improves code maintainability by extracting magic numbers into descriptive named constants throughout the codebase.

Changes:
- Added nanochat/constants.py to centralize commonly used constants
- Replaced hardcoded values for model architecture (aspect ratio, head dimensions, rotary cache multiplier, logit softcap)
- Replaced KV cache growth values with configurable constants using efficient bitwise rounding
- Replaced training hyperparameter defaults (warmup ratios, momentum ramp, EMA decay)
- Replaced calculator timeout with named constant

Benefits:
- Self-documenting code (e.g., ROTARY_CACHE_MULTIPLIER vs 10)
- Easier tuning and experimentation (change values in one place)
- Reduced risk of values drifting out of sync across files
- Preserves all original comments and functionality

No functional changes - all values remain identical to original implementation.

### CODE REVIEW

**Comment by hmms117** [comment]
## Live 35M Engram board update — 2026-09-20

The bounded 12-arm source-placement board is now cleanly running. This is an early-signal comparison: no promotion decision is being made from partial loss.

```text
12/12 full arms running
Puma:  1210676_[0-5]  control, legacy, ds_early, ds_quarter, ds_middle, ds_late
AI:    1210712_6       ds_dual
AI:    1210724_[7-11]  qwen_early, qwen_quarter, qwen_middle, qwen_late, qwen_dual
all:   seed 42, 5k steps, reviewed 35M config, shard-only allocations
```

Infrastructure gates passed before training:

- Puma shard-only CUDA tensor probe `1210665`, `shard:30`, exit 0.
- AI Blackwell CUDA tensor probe `1210683`, `shard:40`, exit 0 with the required non-Hopper path.
- Delta CUDA parity/speed smokes `1210654`, `1210655`, `1210656`, all exit 0.
- Pika is excluded; no jobs were submitted there.

Queue hygiene removed six short screen jobs (`1210706_[6-11]`) and duplicate pending arrays (`1210714`, `1210715`). Those were not ranking evidence and were consuming the AI per-user shard quota. The original full DS-dual task was retained; the five qwen arms were relaunched once as `1210724_[7-11]`. No new arm, seed, or speculative array was added.

Live receipts are present under:

`/z/bio/proj/nanoplm/runs/20260919_engram_source35m_bio12/`

Puma control reached step 2500 with finite `val/aa20 2.7699`, `Delta(aa20-train) +0.0021`, and approximately `279886 tok/s` in-process. AI DS-dual reached step 600 with finite training loss `2.8622` at `78,774,272` tokens. These are progress receipts only; endpoint ranking waits for all 12 arms to reach 5k.

Prior context retained in the board: the matched 181M Engram runs improved mean `val/aa20` by `0.0042` and VTScore by `0.0053` versus baseline across three seeds, with a `+0.0035` generalization-gap change and mixed tiny biological probes. The earlier 80M Delta board kept legacy `delta_ref` as the loss reference; faster candidates did not pass the planned three-seed non-inferiority bound.

PR #74 remains closed and unmerged at `dc87b10db3f2e0d0939f007fcbe75a9d15686acb`. This board update does not mark the PR benchmark-ready. The exact current head still needs independent review, and the known review items remain: one-day Slurm time limit, incomplete cache exports, validation-pair boundary, and batch-schedule/token-accounting handling. No merge or force-push was performed.

**Comment by hmms117** [comment]
## Engram board reconciled; next Puma pair launched

The planned 12-arm 35M source-placement board is complete. All arms reached
5,000 steps / 655.36M tokens with terminal validation, finite loss and zero
skipped batches at exact commit `7f5b3006607e7aeaad58db3547fb1acbabf59645`.

The early-signal leader is `qwen_middle`: Qwen-style source Engram, site `[5]`,
`memory_dim=32`, n-grams `[2,3,4]`, table LR `0.005`, structure disabled.
Terminal metrics were train `2.738873927`, val/aa20 `2.727888821`, Delta
`-0.010985106`, VTScore `2.731184353`. `qwen_quarter` is retained as the
runner-up because it had the best small 3Di top-1 sanity probe. Hardware was
Puma H200 versus AI Blackwell, so speed is not compared; the board is also not
same-capacity (control ~34.69M, 32-dim Engram ~40.18M). This is early-signal,
not promotion evidence.

Existing 181M Engram receipts tested the older generic AA Engram with
`memory_dim=16`; they are not duplicates of the selected source recipe. I
therefore launched one fresh 181M control/Qwen-middle confirmation pair:

```text
array       1210938
control     1210939
treatment   1210938
placement   Puma, low, shard:40
state       both RUNNING
dependency  (null)
source      7f5b3006607e7aeaad58db3547fb1acbabf59645
run root    /z/bio/proj/nanoplm/runs/20260920_engram181m_qwen_middle/
```

The pair shares seed 42, sorted OMG50 auto-EOS order, batch, optimizer,
schedule, 10k-step/81.92M-token horizon and eval config. The meta-model gate
reproduced control `183,011,329` parameters and treatment `188,604,001`; this
is explicitly a recipe confirmation with a `5,592,672` parameter delta, not a
same-capacity claim. Focused source tests passed (`4 passed`), activation
receipts are present, the source worktree is clean, and the launcher was
accepted by Slurm without a dependency. Active `tail35_confirm`/`tail35_extend`
and t-0004 arms were not duplicated.

PR #74 remains closed/unmerged and is not being used as a new code-readiness
claim. Full report and durable snapshot:

- `.herdr-project/nanoplm-t-0006/report.md`
- `docs/snapshot/S1-260920-ENGRAM-SOURCE-35M.md`
- `docs/goals/stage1-next-puma-benchmark/state.yaml`
